shared/chat/blocking/block-modal/index.tsx

```diff
@@ -0,0 +1,283 @@
+          checked={props.includeTranscript}
+        />
+      )}
+    </>
+  )
+}
+
+class BlockModal extends React.PureComponent<Props, State> {
+  state = {
+    blockTeam: true,
+    extraNotes: '',
+    finishClicked: false,
+    includeTranscript: false,
+    // newBlocks holds a Map of blocks that will be applied when user clicks
+    // "Finish" button.
+    newBlocks: new Map(),
+    reportReason: reasons[0],
+    shouldReport: false,
+  }
+
+  componentDidMount() {
+    // Once we get here, trigger actions to refresh current block state of
+    // users.
+    this.props.refreshBlocks()
+
+    // Set default checkbox block values for adder user. We don't care if they
+    // are already blocked, setting a block is idempotent.
+    if (this.props.blockByDefault) {
+      const map = this.state.newBlocks
+      map.set(this.props.adderUsername, {chatBlocked: true, followBlocked: true})
+      this.setState({newBlocks: new Map(map)})
+    }
+  }
+
+  setReport(checked: boolean) {
+    this.setState({shouldReport: checked})
+  }
+
+  setReportReason(reason: string) {
+    this.setState({reportReason: reason})
+  }
+
+  getBlockFor(username: string, which: BlockType) {
+    // First get a current setting from a checkbox, if user has checked anything.
+    const {newBlocks} = this.state
+    const current = newBlocks.get(username)
+    if (current && current[which] !== undefined) {
+      return current[which] || false
+    }
+    // If we don't have a checkbox, check the store for current block value.
+    return this.props.isBlocked(username, which)
+  }
+
+  setBlockFor(username: string, which: BlockType, block: boolean) {
+    const {newBlocks} = this.state
+    const current = newBlocks.get(username)
+    if (current) {
+      current[which] = block
+      newBlocks.set(username, current)
+    } else {
+      newBlocks.set(username, {[which]: block})
+    }
+    // Need to make a new object so the component re-renders.
+    this.setState({newBlocks: new Map(newBlocks)})
+  }
+
+  setBlockTeam(checked: boolean) {
+    this.setState({blockTeam: checked})
+  }
+
+  setExtraNotes(text: string) {
+    this.setState({extraNotes: text})
+  }
+
+  setIncludeTranscript(checked: boolean) {
+    this.setState({includeTranscript: checked})
+  }
+
+  onFinish() {
+    this.props.onFinish(this.state.newBlocks, this.state.blockTeam)
+    this.setState({finishClicked: true})
+  }
+
+  render() {
+    const {teamname, adderUsername} = this.props
+
+    return (
+      <Kb.Modal
+        mode="Default"
+        header={{
+          leftButton: (
+            <Kb.Text onClick={this.props.onCancel} type="BodyPrimaryLink">
+              Cancel
+            </Kb.Text>
+          ),
+          title: <Kb.Icon type="iconfont-block-user" sizeType="Big" color="red" />,
+        }}
+        footer={{
+          content: (
+            <Kb.Button label="Finish" onClick={() => this.onFinish()} fullWidth={true} type="Danger" />
```

Comment:
`onClick={this.onFinish}`, several of these below can be changed too

Reply:
will it bind `this` automatically?

Reply:
iirc it will if you also change the methods to `onFinish = () =>`

Reply:
i think then the methods are created dynamically during construction and bound to the object, so the binding on the call site doesn't matter. I can use this pattern if it's preferred

Reply:
it has the additional benefit of not recreating props that are passed to `Button` on each render.

Reply:
good point. thanks!